more_itertools/more.py

```diff
@@ -1635,6 +1636,24 @@ def islice_extended(iterable, *args):
                 yield item
 
 
+def reversed_extended(iterable):
+    """An extension of :func:`reversed` that supports all iterables, not
+    just those which implement the ``Reversible`` or ``Sequence`` protocols.
+
+        >>> print(*reversed_extended(x for x in range(3)))
+        2 1 0
+
+    If the iterable is already reversible, this function returns the
+    result of :func:`reversed()`. If the iterable is not reversible,
+    this function will cache the remaining items in the iterable and
+    yield them in reverse order, which may require significant storage.
+    """
+    try:
+        return reversed(iterable)
+    except TypeError:
+        return reversed(deque(iterable))
```

Comment:
I think I'd rather see `list` here instead of `deque`. [Both](https://github.com/python/cpython/blob/a5552f023e1d8cbafee1e51d316cc581deb2295f/Modules/_collectionsmodule.c#L1846) [have](https://github.com/python/cpython/blob/1fb72d2ad243c965d4432b4e93884064001a2607/Objects/listobject.c#L2923) special reverse iterator implementations; neither falls back to the "`__len__` and then `__getitem__` in reverse order" default method.

The `deque` would be faster if we were destructively popping through it, of course, but I think profiling should show `list` to be just as fast (and maybe slightly faster due to less accounting) for sequential access.

Reply:
My tests show that there is no difference between the two in Cython, but a clear win for `list()` in PyPy. I'll change it to `list()`.